mlir-sycl/lib/Dialect/SYCL/Transforms/ConstantPropagation.cpp

```diff
@@ -0,0 +1,212 @@
+
+namespace mlir {
+namespace sycl {
+#define GEN_PASS_DEF_CONSTANTPROPAGATIONPASS
+#include "mlir/Dialect/SYCL/Transforms/Passes.h.inc"
+} // namespace sycl
+} // namespace mlir
+
+#define DEBUG_TYPE "sycl-cpp"
+
+using namespace mlir;
+using namespace mlir::sycl;
+
+static raw_ostream &operator<<(raw_ostream &os, const llvm::BitVector &bv) {
+  os << "{";
+  SmallVector<unsigned> set;
+  for (unsigned i = 0, size = static_cast<unsigned>(bv.size()); i < size; ++i) {
+    if (bv.test(i))
+      set.push_back(i);
+  }
+  llvm::interleaveComma(set, os);
+  return os << "}";
+}
+
+namespace {
+class ConstantPropagationPass
+    : public mlir::sycl::impl::ConstantPropagationPassBase<
+          ConstantPropagationPass> {
+public:
+  using ConstantPropagationPassBase<
+      ConstantPropagationPass>::ConstantPropagationPassBase;
+
+  void runOnOperation() final;
+
+private:
+  /// Return a range with all of the constant arguments of \p op.
+  static auto getConstantArgs(SYCLHostScheduleKernel op);
+
+  /// Propagate constants in \p range to the function launched by \p launch.
+  template <typename RangeTy>
+  void propagateConstantArgs(RangeTy constants, SYCLHostScheduleKernel launch);
+
+  /// Drop arguments \p constants from \p launch.
+  template <typename RangeTy>
+  void dropConstantArgs(RangeTy constants, Builder &builder,
+                        SYCLHostScheduleKernel launch);
+};
+
+/// Class representing a constant explicit argument, i.e., those kernel
+/// functions receive as a parameter.
+class ConstantExplicitArg {
+public:
+  enum class Kind { ConstantArithArg };
+
+  unsigned getIndex() const { return index; }
+  Kind getKind() const { return kind; }
+
+protected:
+  ConstantExplicitArg(Kind kind, unsigned index) : kind(kind), index(index) {}
+
+private:
+  Kind kind;
+  unsigned index;
+};
+
+/// Class representing a constant arithmetic argument.
+class ConstantArithArg : public ConstantExplicitArg {
+public:
+  ConstantArithArg(unsigned index, Operation *definingOp)
+      : ConstantExplicitArg(ConstantExplicitArg::Kind::ConstantArithArg, index),
+        definingOp(definingOp) {
+    assert(definingOp && "Expecting valid operation");
+    assert(definingOp->hasTrait<OpTrait::ConstantLike>() &&
+           "Expecting constant");
+  }
+
+  /// Propagate the constant this argument represents.
+  void propagate(OpBuilder builder, Region &region) const;
+
+  static bool classof(const ConstantExplicitArg *c) {
+    return c->getKind() == ConstantExplicitArg::Kind::ConstantArithArg;
+  }
+
+private:
+  Operation *definingOp;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// ConstantPropagationPass
+//===----------------------------------------------------------------------===//
+
+void ConstantArithArg::propagate(OpBuilder builder, Region &region) const {
+  Value toReplace = region.getArgument(getIndex());
+  toReplace.replaceAllUsesWith(builder.clone(*definingOp)->getResult(0));
```

Comment:
Is it safe to use the original constant operation or do we need to duplicate it at the beginning of the GPU/kernel function?

Reply:
I'm `builder.clone`-ing there